crates/oci-distribution/src/client.rs

```diff
@@ -114,6 +127,73 @@ impl Client {
         })
     }
 
+    /// Push an image and return the uploaded URL of the image
+    ///
+    /// The client will check if it's already been authenticated and if
+    /// not will attempt to do.
+    ///
+    /// If a manifest is not provided, the client will attempt to generate
+    /// it from the provided image and config data.
+    ///
+    /// Returns pullable URL for the image
+    pub async fn push(
+        &mut self,
+        image_ref: &Reference,
+        image_data: &ImageData,
+        config_data: &[u8],
+        config_media_type: &str,
+        auth: &RegistryAuth,
+        manifest: Option<OciManifest>,
+    ) -> anyhow::Result<String> {
+        debug!("Pushing image: {:?}", image_ref);
+
+        if !self.tokens.contains_key(image_ref.registry()) {
+            self.auth(image_ref, auth, &RegistryOperation::Push).await?;
+        }
+
+        // Start push session
+        let mut location = self.begin_push_session(image_ref).await?;
+
+        // Upload layers
+        let mut start_byte = 0;
+        for layer in &image_data.layers {
+            // Destructuring assignment is not yet supported
+            let (next_location, next_byte) = self
+                .push_layer(&location, &image_ref, layer.data.to_vec(), start_byte)
+                .await?;
+            location = next_location;
+            start_byte = next_byte;
+        }
+
+        // End push session, upload manifest
+        let image_url = self
+            .end_push_session(
+                &location,
+                &image_ref,
+                &sha256_digest(
+                    &image_data
```

Comment:
Something else occurs to me here, ImageData also includes a `digest` field, which is an optional field. I think I'll make a method that either returns the value in the field (if it has been provided) or computes it otherwise.